src/basic/recurse-dir.c

```diff
@@ -84,6 +85,12 @@ int readdir_all(int dir_fd,
                 if (ignore_dirent(entry, flags))
                         continue;
 
+                if (FLAGS_SET(flags, RECURSE_DIR_ENSURE_TYPE)) {
+                        r = dirent_ensure_type(dir_fd, entry);
+                        if (r < 0 && r != -ENOENT)
+                                return r;
```

Comment:
so, hmm, question is what the best approach to handle ENOENT here is. Right now you just leve those dentry with DT_UNKNOWN. I think it would probably be better to suppress them altogether. I mean, they don#t exist anymore, so we might as well remove them from the list.

Specifically, maybe in the second FOREACH_DIRENT_IN_BUFFER() loop, below, if RECURSE_DIR_ENSURE_TYPE is set, and we see DT_UNKNOWN still in the dentry, we know it's because it vanished and we saw ENOENT, hence simply skip over the entry so that `de->entries` won't have it (though it remains in `de->buffer`).

net result would be that setting RECURSE_DIR_ENSURE_TYPE gives you the guarantee that d_type is fully initialized, and never DT_UNKNOWN. WHich I think is a good thing for robustness.